franka_example_controllers/src/cartesian_impedance_example_controller.cpp

```diff
@@ -0,0 +1,249 @@
+  position_d_target_ = initial_transform.translation();
+  orientation_d_target_ = initial_transform.rotation();
+
+  q_d_nullspace_ = q_initial;
+  return CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn CartesianImpedanceExampleController::on_deactivate(
+    const rclcpp_lifecycle::State& /*previous_state*/) {
+  franka_robot_state_->release_interfaces();
+  franka_robot_model_->release_interfaces();
+  return CallbackReturn::SUCCESS;
+}
+
+controller_interface::return_type CartesianImpedanceExampleController::update(
+    const rclcpp::Time& time,
+    const rclcpp::Duration& period) {
+  // get state variables
+  robot_state_ = franka_msgs::msg::FrankaRobotState();
+  franka_robot_state_->get_values_as_message(robot_state_);
+
+  std::array<double, 49> mass = franka_robot_model_->getMassMatrix();
+  std::array<double, 7> coriolis_array = franka_robot_model_->getCoriolisForceVector();
+  std::array<double, 42> jacobian_array =
+      franka_robot_model_->getZeroJacobian(franka::Frame::kEndEffector);
+
+  Eigen::Map<Eigen::Matrix<double, 7, 7>> M(mass.data());
+  Eigen::Map<Eigen::Matrix<double, 7, 1>> coriolis(coriolis_array.data());
+  Eigen::Map<Eigen::Matrix<double, 6, 7>> jacobian(jacobian_array.data());
+
+  Eigen::Map<Eigen::Matrix<double, 7, 1>> q(robot_state_.measured_joint_state.position.data());
+  Eigen::Map<Eigen::Matrix<double, 7, 1>> dq(robot_state_.measured_joint_state.velocity.data());
+  Eigen::Map<Eigen::Matrix<double, 7, 1>> tau_J_d(robot_state_.measured_joint_state.effort.data());
+  Eigen::Vector3d position(robot_state_.o_t_ee.pose.position.x, robot_state_.o_t_ee.pose.position.y,
+                           robot_state_.o_t_ee.pose.position.z);
+  Eigen::Quaterniond orientation(
+      robot_state_.o_t_ee.pose.orientation.w, robot_state_.o_t_ee.pose.orientation.x,
+      robot_state_.o_t_ee.pose.orientation.y, robot_state_.o_t_ee.pose.orientation.z);
+  Eigen::Affine3d transform = Eigen::Affine3d::Identity();
+  transform.translation() = position;
+  transform.rotate(orientation.toRotationMatrix());
+
+  // compute error to desired pose
+  // position error
+  Eigen::Matrix<double, 6, 1> position_error;
+  position_error.head(3) << position - position_d_;
+
+  // orientation error
+  if (orientation_d_.coeffs().dot(orientation.coeffs()) < 0.0) {
+    orientation.coeffs() << -orientation.coeffs();
+  }
+
+  // "difference" quaternion
+  Eigen::Quaterniond error_quaternion(orientation.inverse() * orientation_d_);
+  position_error.tail(3) << error_quaternion.x(), error_quaternion.y(), error_quaternion.z();
+  // Transform to base frame
+  position_error.tail(3) << -transform.rotation() * position_error.tail(3);
+
+  // compute control
+  // allocate variables
+  Eigen::VectorXd tau_task(7), tau_nullspace(7), tau_d(7);
+
+  // pseudoinverse for nullspace handling
+  Eigen::MatrixXd jacobian_transpose_pinv;
+  pseudoInverse(jacobian.transpose(), jacobian_transpose_pinv);
```

Comment:
based on the definition on PseudoInverse from you:
```
inline Eigen::MatrixXd pseudoInverse(const Eigen::MatrixXd& M_, bool damped = true) {
  double lambda_ = damped ? 0.2 : 0.0;

  Eigen::JacobiSVD<Eigen::MatrixXd> svd(M_, Eigen::ComputeFullU | Eigen::ComputeFullV);
  Eigen::JacobiSVD<Eigen::MatrixXd>::SingularValuesType sing_vals_ = svd.singularValues();
  Eigen::MatrixXd S_ = M_;  // copying the dimensions of M_, its content is not needed.
  S_.setZero();

  for (int i = 0; i < sing_vals_.size(); i++)
    S_(i, i) = (sing_vals_(i)) / (sing_vals_(i) * sing_vals_(i) + lambda_ * lambda_);

  return svd.matrixV() * S_.transpose() * svd.matrixU().transpose();
}

```
second parameter should be `bool damped`. And the returned value should be the inverted matrix.